### PR TITLE
Use a subset of an addon’s package for generating its cache key

### PR DESCRIPTION
**Pull request by dgeb**
`calculateCacheKeyForTree` has been using the entirety of an addon’s `package.json` parsed contents (available as `addonInstance.pkg`) in generating a unique cache key.

The package spec has has reserved properties with `_` and `$` prefixes for package registries to use at their discretion (see http://wiki.commonjs.org/wiki/Packages/1.1). `npm` has begun to take advantage of this by embedding properties that are completely unique to the package (such as its parent). This of course defeats the entire purpose of the cache key.

Therefore, this change moves the implementation to only rely on certain properties that signal the uniqueness of an addon: its version, name, and dependencies (of all forms). This avoids the potential problems with private properties, and should be a bit faster to hash as well.

### CODE REVIEW

File: lib/calculate-cache-key-for-tree.js
Comment (dgeb) [inline lib/calculate-cache-key-for-tree.js]: I am not sure if `name` (from `addonInstance.pkg.name`) is always entirely redundant with `addonInstance.name`, so I cautiously left both.

Comment (rwjblue) [reply lib/calculate-cache-key-for-tree.js]: Your caution has paid off 😸, they can indeed be different (though we are pushing very hard to ensure that this is "going away").

File: lib/calculate-cache-key-for-tree.js
Comment (rwjblue) [inline lib/calculate-cache-key-for-tree.js]: 😭 - This package still supports Node 4, so we can't use destructuring just yet. 

Mind changing to basically:

```js
  let cacheKeyParts = [
    addonInstance.name,
    treeType,
    pkg.name,
    pkg.version,
    pkg.dependencies,
    pkg.devDependencies,
    pkg.peerDependencies,
    pkg.bundledDependencies,
    pkg.optionalDependencies
  ].concat(additionalCacheKeyParts);
```

---

I'm happy to drop Node 4 support generally speaking, but I'd love getting this bugfix in before doing that...

**Review by rwjblue — CHANGES_REQUESTED** [review]

File: lib/calculate-cache-key-for-tree.js
Comment (dgeb) [reply lib/calculate-cache-key-for-tree.js]: Sure no prob! Will be good to drop Node 4 (and 6 for that matter), but I agree it's better to do in a separate step.

**Review by stefanpenner — APPROVED** [review]

File: lib/calculate-cache-key-for-tree.js
Comment (stefanpenner) [reply lib/calculate-cache-key-for-tree.js]: Ya, let's do that after this lands.

**Comment by stefanpenner** [comment]
@dgeb  lgtm, once green we can merge/release.

**Comment by stefanpenner** [comment]
@dgeb any additional work planned?

**Comment by dgeb** [comment]
> any additional work planned?

@stefanpenner No more work planned in this package, so let's get a release out after it's green.

I do have some related work happening in https://github.com/dgeb/ember-cli-addon-guard as well as a separate addon for version-based namespacing (see the readme of ember-cli-addon-guard). I've been working with @rwjblue on all of this but would appreciate your feedback as well.

**Comment by stefanpenner** [comment]
released as v1.2.0 🎉

**Comment by dgeb** [comment]
thanks @stefanpenner 💯

**Pull request merged**
